pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,345 @@
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("nodemetrics").With("node", req.Name))
+	// Remove the previous gauge after node labels are updated
+	c.deleteGauges(req.NamespacedName)
+	// Retrieve node from reconcile request
+	node := &v1.Node{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	if err := c.updateGauges(ctx, node); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to update gauges: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	return controllerruntime.
+		NewControllerManagedBy(m).
+		Named("nodemetrics").
+		For(&v1.Node{}).
+		Watches(
+			// Reconcile all nodes related to a provisioner when it changes.
+			&source.Kind{Type: &v1alpha5.Provisioner{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				nodes := &v1.NodeList{}
+				if err := c.KubeClient.List(ctx, nodes, client.MatchingLabels(map[string]string{v1alpha5.ProvisionerNameLabelKey: o.GetName()})); err != nil {
+					logging.FromContext(ctx).Errorf("Failed to list nodes when mapping expiration watch events, %s", err.Error())
+					return requests
+				}
+				for _, node := range nodes.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}})
+				}
+				return requests
+			}),
+		).
+		Watches(
+			// Reconcile nodes where pods have changed
+			&source.Kind{Type: &v1.Pod{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				if name := o.(*v1.Pod).Spec.NodeName; name != "" {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: name}})
+				}
+				return requests
+			}),
+		).
+		Complete(c)
+}
+
+func (c *Controller) deleteGauges(nodeNamespacedName types.NamespacedName) {
+	if labelSlice, ok := c.LabelSliceMap[nodeNamespacedName]; ok {
+		for _, labels := range labelSlice {
+			allocatableGaugeVec.Delete(labels)
+			podRequestsGaugeVec.Delete(labels)
+			podLimitsGaugeVec.Delete(labels)
+			daemonRequestsGaugeVec.Delete(labels)
+			daemonLimitsGaugeVec.Delete(labels)
+			overheadGaugeVec.Delete(labels)
+		}
+	}
+	c.LabelSliceMap[nodeNamespacedName] = []prometheus.Labels{}
+
+}
+
+// generateLabels creates the labels using the current state of the pod
+func (c *Controller) generateLabels(node *v1.Node, resourceTypeName string) prometheus.Labels {
+	metricLabels := prometheus.Labels{}
+	metricLabels[resourceType] = resourceTypeName
+	metricLabels[nodeName] = node.GetName()
+	if provisionerName, ok := node.Labels[v1alpha5.ProvisionerNameLabelKey]; !ok {
+		metricLabels[nodeProvisioner] = "N/A"
```

Comment:
Why does this have to be N/A? Can't we just use "" and simplify to 

metricsLabels[nodeProvisioner] = node.Labels[v1alpha5.ProvisionerNameLabelKey]

Reply:
I tried this. Unfortunately Prometheus automatically ignores the label value. It creates confusion since we can no longer distinguish if the label value doesn't exist or it is just empty.

Also this N/A is more human readable so that the customer will know that we don't have a value for this label.

Reply:
It would be nice to just let them be empty string, and then we can postprocess later. This isolates prometheus's quirk and allows us to follow go conventions elsewhere.
```
for key, value := range metricLabels {
  if value == "" {
    metricLabels[key] = "N/A"
  }
}
```